cmd/gen-key/ecdsa.go

```diff
@@ -0,0 +1,212 @@
+				// PKCS#11 v2.20 specified that the CKA_EC_POINT was to be stored in a DER-encoded
+				// OCTET STRING.
+				var point asn1.RawValue
+				_, err = asn1.Unmarshal(a.Value, &point)
+				if err != nil {
+					return nil, fmt.Errorf("Failed to unmarshal returned CKA_EC_POINT: %s", err)
+				}
+				if len(point.Bytes) == 0 {
+					return nil, errors.New("Invalid CKA_EC_POINT value returned, OCTET string is empty")
+				}
+				x, y = elliptic.Unmarshal(expectedCurve, point.Bytes)
+				if x == nil {
+					fmt.Println(point.Bytes)
+					return nil, errors.New("Invalid CKA_EC_POINT value returned, point is malformed")
+				}
+			}
+			pubKey.X, pubKey.Y = x, y
+			gotPoint = true
+			log.Printf("\tX: %X\n", pubKey.X.Bytes())
+			log.Printf("\tY: %X\n", pubKey.Y.Bytes())
+		}
+	}
+	if !gotPoint || !gotCurve {
+		return nil, errors.New("Couldn't retrieve EC point and EC parameters")
+	}
+	return pubKey, nil
+}
+
+// ecVerify verifies that the extracted public key corresponds with the generated
+// private key on the device, specified by the provided object handle, by signing
+// a nonce generated on the device and verifying the returned signature using the
+// public key.
+func ecVerify(ctx PKCtx, session pkcs11.SessionHandle, object pkcs11.ObjectHandle, pub *ecdsa.PublicKey) error {
+	err := ctx.SignInit(session, []*pkcs11.Mechanism{pkcs11.NewMechanism(pkcs11.CKM_ECDSA, nil)}, object)
+	if err != nil {
+		return fmt.Errorf("failed to initialize signing operation: %s", err)
+	}
+	nonce, err := getRandomBytes(ctx, session)
+	if err != nil {
+		return fmt.Errorf("failed to construct nonce: %s", err)
+	}
+	log.Printf("\tConstructed nonce: %d (%X)\n", big.NewInt(0).SetBytes(nonce), nonce)
+	hashFunc := curveToHash[pub.Curve.Params()].New()
+	hashFunc.Write(nonce)
+	hash := hashFunc.Sum(nil)
+	log.Printf("\tMessage %s hash: %X\n", hashToString[curveToHash[pub.Curve.Params()]], hash)
+	signature, err := ctx.Sign(session, hash)
+	if err != nil {
+		return fmt.Errorf("failed to sign data: %s", err)
+	}
+	log.Printf("\tMessage signature: %X\n", signature)
+	r := big.NewInt(0).SetBytes(signature[:len(signature)/2])
+	s := big.NewInt(0).SetBytes(signature[len(signature)/2:])
+	if !ecdsa.Verify(pub, hash[:], r, s) {
+		return errors.New("failed to verify ECDSA signature over test data")
+	}
+	log.Println("\tSignature verified")
+	return nil
+}
+
+// ecGenerate is used to generate and verify a ECDSA key pair of the type
+// specified by curveStr and with the provided label. For devices that implement
+// a pre-2.11 version of the PKCS#11 specification compatMode should be true.
+// It returns the public part of the generated key pair as a ecdsa.PublicKey.
+func ecGenerate(ctx PKCtx, session pkcs11.SessionHandle, label, curveStr string, compatMode bool) (*ecdsa.PublicKey, error) {
+	curve, present := stringToCurve[curveStr]
+	if !present {
+		return nil, errors.New("curve not supported")
```

Comment:
tiny nit: It might be slightly more user friendly to have the `errors.New("curve not supported")` replaced with `fmt.Errorf("curve %q not supported", curveStr)`.